### ISSUE TITLE
wikizimmer http error handling needs improvement

### ISSUE THREAD

**Issue by fattredd** [vss-devel/zimmer#3]
I've got wikizimmer running into a 403 error on an image request. It chokes and stops the whole process when it encounters this.

Here's the command I'm running to generate the error log below:
`wikizimmer https://www.dandwiki.com 2> error.log`

```log
Save error Error Load error 404 https://www.dandwiki.com/w/images/searchgray_iconsmall.png //www.dandwiki.com/w/images/searchgray_iconsmall.png -> /-/searchgray_iconsmall.png
Save error Error Load error 404 https://www.dandwiki.com/w/images/search_icon.png //www.dandwiki.com/w/images/search_icon.png -> /-/search_icon.png
Save error Error Load error 404 https://www.dandwiki.com/w/images/browse_iconsmall.png //www.dandwiki.com/w/images/browse_iconsmall.png -> /-/browse_iconsmall.png
Save error Error Load error 404 https://www.dandwiki.com/w/images/browse_icon.png //www.dandwiki.com/w/images/browse_icon.png -> /-/browse_icon.png
Save error Error Load error 404 https://www.dandwiki.com/w/images/world.png //www.dandwiki.com/w/images/world.png -> /-/world.png
Save error Error Load error 404 https://www.dandwiki.com/w/images/rdf_flyer.18.gif //www.dandwiki.com/w/images/rdf_flyer.18.gif -> /-/rdf_flyer.18.gif
Save error Error Load error 404 https://www.dandwiki.com/w/world.png //www.dandwiki.com/w/world.png -> /-/world.png
Save error Error Load error 404 https://www.dandwiki.com/w/rdf_flyer.18.gif //www.dandwiki.com/w/rdf_flyer.18.gif -> /-/rdf_flyer.18.gif
request fatal error { StatusCodeError: 403 - {"type":"Buffer","data":[60,33,68,79,67,84,89,80,69,32,104,116,109,108,62,10,60,104,116,109,108,32,115,116,121,108,101,61,34,104,101,105,103,104,116,58,49,48,48,37,34,62,10,60,104,101,97,100,62,60,116,105,116,108,101,62,32,52,48,51,32,70,111,114,98,105,100,100,101,110,13,10,60,47,116,105,116,108,101,62,60,47,104,101,97,100,62,10,60,98,111,100,121,32,115,116,121,108,101,61,34,99,111,108,111,114,58,32,35,52,52,52,59,32,109,97,114,103,105,110,58,48,59,102,111,110,116,58,32,110,111,114,109,97,108,32,49,52,112,120,47,50,48,112,120,32,65,114,105,97,108,44,32,72,101,108,118,101,116,105,99,97,44,32,115,97,110,115,45,115,101,114,105,102,59,32,104,101,105,103,104,116,58,49,48,48,37,59,32,98,97,99,107,103,114,111,117,110,100,45,99,111,108,111,114,58,32,35,102,102,102,59,34,62,10,60,100,105,118,32,115,116,121,108,101,61,34,104,101,105,103,104,116,58,97,117,116,111,59,32,109,105,110,45,104,101,105,103,104,116,58,49,48,48,37,59,32,34,62,32,32,32,32,32,60,100,105,118,32,115,116,121,108,101,61,34,116,101,120,116,45,97,108,105,103,110,58,32,99,101,110,116,101,114,59,32,119,105,100,116,104,58,56,48,48,112,120,59,32,109,97,114,103,105,110,45,108,101,102,116,58,32,45,52,48,48,112,120,59,32,112,111,115,105,116,105,111,110,58,97,98,115,111,108,117,116,101,59,32,116,111,112,58,32,51,48,37,59,32,108,101,102,116,58,53,48,37,59,34,62,10,32,32,32,32,32,32,32,32,60,104,49,32,115,116,121,108,101,61,34,109,97,114,103,105,110,58,48,59,32,102,111,110,116,45,115,105,122,101,58,49,53,48,112,120,59,32,108,105,110,101,45,104,101,105,103,104,116,58,49,53,48,112,120,59,32,102,111,110,116,45,119,101,105,103,104,116,58,98,111,108,100,59,34,62,52,48,51,60,47,104,49,62,10,60,104,50,32,115,116,121,108,101,61,34,109,97,114,103,105,110,45,116,111,112,58,50,48,112,120,59,102,111,110,116,45,115,105,122,101,58,32,51,48,112,120,59,34,62,70,111,114,98,105,100,100,101,110,13,10,60,47,104,50,62,10,60,112,62,65,99,99,101,115,115,32,116,111,32,116,104,105,115,32,114,101,115,111,117,114,99,101,32,111,110,32,116,104,101,32,115,101,114,118,101,114,32,105,115,32,100,101,110,105,101,100,33,60,47,112,62,10,60,47,100,105,118,62,60,47,100,105,118,62,60,100,105,118,32,115,116,121,108,101,61,34,99,111,108,111,114,58,35,102,48,102,48,102,48,59,32,102,111,110,116,45,115,105,122,101,58,49,50,112,120,59,109,97,114,103,105,110,58,97,117,116,111,59,112,97,100,100,105,110,103,58,48,112,120,32,51,48,112,120,32,48,112,120,32,51,48,112,120,59,112,111,115,105,116,105,111,110,58,114,101,108,97,116,105,118,101,59,99,108,101,97,114,58,98,111,116,104,59,104,101,105,103,104,116,58,49,48,48,112,120,59,109,97,114,103,105,110,45,116,111,112,58,45,49,48,49,112,120,59,98,97,99,107,103,114,111,117,110,100,45,99,111,108,111,114,58,35,52,55,52,55,52,55,59,98,111,114,100,101,114,45,116,111,112,58,32,49,112,120,32,115,111,108,105,100,32,114,103,98,97,40,48,44,48,44,48,44,48,46,49,53,41,59,98,111,120,45,115,104,97,100,111,119,58,32,48,32,49,112,120,32,48,32,114,103,98,97,40,50,53,53,44,32,50,53,53,44,32,50,53,53,44,32,48,46,51,41,32,105,110,115,101,116,59,34,62,10,60,98,114,62,80,114,111,117,100,108,121,32,112,111,119,101,114,101,100,32,98,121,32,32,60,97,32,115,116,121,108,101,61,34,99,111,108,111,114,58,35,102,102,102,59,34,32,104,114,101,102,61,34,104,116,116,112,58,47,47,119,119,119,46,108,105,116,101,115,112,101,101,100,116,101,99,104,46,99,111,109,47,101,114,114,111,114,45,112,97,103,101,34,62,76,105,116,101,83,112,101,101,100,32,87,101,98,32,83,101,114,118,101,114,60,47,97,62,60,112,62,80,108,101,97,115,101,32,98,101,32,97,100,118,105,115,101,100,32,116,104,97,116,32,76,105,116,101,83,112,101,101,100,32,84,101,99,104,110,111,108,111,103,105,101,115,32,73,110,99,46,32,105,115,32,110,111,116,32,97,32,119,101,98,32,104,111,115,116,105,110,103,32,99,111,109,112,97,110,121,32,97,110,100,44,32,97,115,32,115,117,99,104,44,32,104,97,115,32,110,111,32,99,111,110,116,114,111,108,32,111,118,101,114,32,99,111,110,116,101,110,116,32,102,111,117,110,100,32,111,110,32,116,104,105,115,32,115,105,116,101,46,60,47,112,62,60,47,100,105,118,62,60,47,98,111,100,121,62,60,47,104,116,109,108,62,10]}
    at new StatusCodeError (/usr/lib/node_modules/zimmer/node_modules/request-promise-core/lib/errors.js:32:15)
    at Request.plumbing.callback (/usr/lib/node_modules/zimmer/node_modules/request-promise-core/lib/plumbing.js:104:33)
    at Request.RP$callback [as _callback] (/usr/lib/node_modules/zimmer/node_modules/request-promise-core/lib/plumbing.js:46:31)
    at Request.self.callback (/usr/lib/node_modules/zimmer/node_modules/request/request.js:185:22)
    at emitTwo (events.js:106:13)
    at Request.emit (events.js:191:7)
    at Request.<anonymous> (/usr/lib/node_modules/zimmer/node_modules/request/request.js:1157:10)
    at emitOne (events.js:96:13)
    at Request.emit (events.js:188:7)
    at IncomingMessage.<anonymous> (/usr/lib/node_modules/zimmer/node_modules/request/request.js:1079:12)
    at IncomingMessage.g (events.js:292:16)
    at emitNone (events.js:91:20)
    at IncomingMessage.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:978:12)
    at _combinedTickCallback (internal/process/next_tick.js:80:11)
    at process._tickCallback (internal/process/next_tick.js:104:9)
  name: 'StatusCodeError',
  statusCode: 403,
  message: '403 - {"type":"Buffer","data":[60,33,68,79,67,84,89,80,69,32,104,116,109,108,62,10,60,104,116,109,108,32,115,116,121,108,101,61,34,104,101,105,103,104,116,58,49,48,48,37,34,62,10,60,104,101,97,100,62,60,116,105,116,108,101,62,32,52,48,51,32,70,111,114,98,105,100,100,101,110,13,10,60,47,116,105,116,108,101,62,60,47,104,101,97,100,62,10,60,98,111,100,121,32,115,116,121,108,101,61,34,99,111,108,111,114,58,32,35,52,52,52,59,32,109,97,114,103,105,110,58,48,59,102,111,110,116,58,32,110,111,114,109,97,108,32,49,52,112,120,47,50,48,112,120,32,65,114,105,97,108,44,32,72,101,108,118,101,116,105,99,97,44,32,115,97,110,115,45,115,101,114,105,102,59,32,104,101,105,103,104,116,58,49,48,48,37,59,32,98,97,99,107,103,114,111,117,110,100,45,99,111,108,111,114,58,32,35,102,102,102,59,34,62,10,60,100,105,118,32,115,116,121,108,101,61,34,104,101,105,103,104,116,58,97,117,116,111,59,32,109,105,110,45,104,101,105,103,104,116,58,49,48,48,37,59,32,34,62,32,32,32,32,32,60,100,105,118,32,115,116,121,108,101,61,34,116,101,120,116,45,97,108,105,103,110,58,32,99,101,110,116,101,114,59,32,119,105,100,116,104,58,56,48,48,112,120,59,32,109,97,114,103,105,110,45,108,101,102,116,58,32,45,52,48,48,112,120,59,32,112,111,115,105,116,105,111,110,58,97,98,115,111,108,117,116,101,59,32,116,111,112,58,32,51,48,37,59,32,108,101,102,116,58,53,48,37,59,34,62,10,32,32,32,32,32,32,32,32,60,104,49,32,115,116,121,108,101,61,34,109,97,114,103,105,110,58,48,59,32,102,111,110,116,45,115,105,122,101,58,49,53,48,112,120,59,32,108,105,110,101,45,104,101,105,103,104,116,58,49,53,48,112,120,59,32,102,111,110,116,45,119,101,105,103,104,116,58,98,111,108,100,59,34,62,52,48,51,60,47,104,49,62,10,60,104,50,32,115,116,121,108,101,61,34,109,97,114,103,105,110,45,116,111,112,58,50,48,112,120,59,102,111,110,116,45,115,105,122,101,58,32,51,48,112,120,59,34,62,70,111,114,98,105,100,100,101,110,13,10,60,47,104,50,62,10,60,112,62,65,99,99,101,115,115,32,116,111,32,116,104,105,115,32,114,101,115,111,117,114,99,101,32,111,110,32,116,104,101,32,115,101,114,118,101,114,32,105,115,32,100,101,110,105,101,100,33,60,47,112,62,10,60,47,100,105,118,62,60,47,100,105,118,62,60,100,105,118,32,115,116,121,108,101,61,34,99,111,108,111,114,58,35,102,48,102,48,102,48,59,32,102,111,110,116,45,115,105,122,101,58,49,50,112,120,59,109,97,114,103,105,110,58,97,117,116,111,59,112,97,100,100,105,110,103,58,48,112,120,32,51,48,112,120,32,48,112,120,32,51,48,112,120,59,112,111,115,105,116,105,111,110,58,114,101,108,97,116,105,118,101,59,99,108,101,97,114,58,98,111,116,104,59,104,101,105,103,104,116,58,49,48,48,112,120,59,109,97,114,103,105,110,45,116,111,112,58,45,49,48,49,112,120,59,98,97,99,107,103,114,111,117,110,100,45,99,111,108,111,114,58,35,52,55,52,55,52,55,59,98,111,114,100,101,114,45,116,111,112,58,32,49,112,120,32,115,111,108,105,100,32,114,103,98,97,40,48,44,48,44,48,44,48,46,49,53,41,59,98,111,120,45,115,104,97,100,111,119,58,32,48,32,49,112,120,32,48,32,114,103,98,97,40,50,53,53,44,32,50,53,53,44,32,50,53,53,44,32,48,46,51,41,32,105,110,115,101,116,59,34,62,10,60,98,114,62,80,114,111,117,100,108,121,32,112,111,119,101,114,101,100,32,98,121,32,32,60,97,32,115,116,121,108,101,61,34,99,111,108,111,114,58,35,102,102,102,59,34,32,104,114,101,102,61,34,104,116,116,112,58,47,47,119,119,119,46,108,105,116,101,115,112,101,101,100,116,101,99,104,46,99,111,109,47,101,114,114,111,114,45,112,97,103,101,34,62,76,105,116,101,83,112,101,101,100,32,87,101,98,32,83,101,114,118,101,114,60,47,97,62,60,112,62,80,108,101,97,115,101,32,98,101,32,97,100,118,105,115,101,100,32,116,104,97,116,32,76,105,116,101,83,112,101,101,100,32,84,101,99,104,110,111,108,111,103,105,101,115,32,73,110,99,46,32,105,115,32,110,111,116,32,97,32,119,101,98,32,104,111,115,116,105,110,103,32,99,111,109,112,97,110,121,32,97,110,100,44,32,97,115,32,115,117,99,104,44,32,104,97,115,32,110,111,32,99,111,110,116,114,111,108,32,111,118,101,114,32,99,111,110,116,101,110,116,32,102,111,117,110,100,32,111,110,32,116,104,105,115,32,115,105,116,101,46,60,47,112,62,60,47,100,105,118,62,60,47,98,111,100,121,62,60,47,104,116,109,108,62,10]}',
  error: <Buffer 3c 21 44 4f 43 54 59 50 45 20 68 74 6d 6c 3e 0a 3c 68 74 6d 6c 20 73 74 79 6c 65 3d 22 68 65 69 67 68 74 3a 31 30 30 25 22 3e 0a 3c 68 65 61 64 3e 3c ... >,
  options: 
   { url: 'http://raptureorwrath.net/wp-content/uploads/2013/09/stinging-locusts1-300x240.jpg',
     encoding: null,
     uri: null,
     headers: { 'User-Agent': 'wikizimmer/0.1.0 (https://github.com/vadp/zimmer email:vadp.devl@gmail.com)' },
     resolveWithFullResponse: true,
     timeout: 300000,
     forever: true,
     resolve: [Function],
     reject: [Function],
     retries: 0,
     callback: [Function: RP$callback],
     transform: undefined,
     simple: true,
     transform2xxOnly: false },
  response: 
   IncomingMessage {
     _readableState: 
      ReadableState {
        objectMode: false,
        highWaterMark: 16384,
        buffer: [Object],
        length: 0,
        pipes: null,
        pipesCount: 0,
        flowing: true,
        ended: true,
        endEmitted: true,
        reading: false,
        sync: true,
        needReadable: false,
        emittedReadable: false,
        readableListening: false,
        resumeScheduled: false,
        defaultEncoding: 'utf8',
        ranOut: false,
        awaitDrain: 0,
        readingMore: false,
        decoder: null,
        encoding: null },
     readable: false,
     domain: null,
     _events: 
      { end: [Object],
        close: [Object],
        data: [Function],
        error: [Function] },
     _eventsCount: 4,
     _maxListeners: undefined,
     socket: 
      Socket {
        connecting: false,
        _hadError: false,
        _handle: [Object],
        _parent: null,
        _host: 'raptureorwrath.net',
        _readableState: [Object],
        readable: true,
        domain: null,
        _events: [Object],
        _eventsCount: 8,
        _maxListeners: undefined,
        _writableState: [Object],
        writable: true,
        allowHalfOpen: false,
        destroyed: false,
        _bytesDispatched: 213,
        _sockname: null,
        _pendingData: null,
        _pendingEncoding: '',
        server: null,
        _server: null,
        parser: null,
        _httpMessage: null,
        read: [Function],
        _consuming: true,
        _idleTimeout: -1,
        _idleNext: null,
        _idlePrev: null,
        _idleStart: 1213959 },
     connection: 
      Socket {
        connecting: false,
        _hadError: false,
        _handle: [Object],
        _parent: null,
        _host: 'raptureorwrath.net',
        _readableState: [Object],
        readable: true,
        domain: null,
        _events: [Object],
        _eventsCount: 8,
        _maxListeners: undefined,
        _writableState: [Object],
        writable: true,
        allowHalfOpen: false,
        destroyed: false,
        _bytesDispatched: 213,
        _sockname: null,
        _pendingData: null,
        _pendingEncoding: '',
        server: null,
        _server: null,
        parser: null,
        _httpMessage: null,
        read: [Function],
        _consuming: true,
        _idleTimeout: -1,
        _idleNext: null,
        _idlePrev: null,
        _idleStart: 1213959 },
     httpVersionMajor: 1,
     httpVersionMinor: 1,
     httpVersion: '1.1',
     complete: true,
     headers: 
      { 'cache-control': 'private, no-cache, no-store, must-revalidate, max-age=0',
        pragma: 'no-cache',
        'content-type': 'text/html',
        'content-length': '1139',
        date: 'Mon, 09 Jul 2018 13:22:52 GMT',
        'accept-ranges': 'bytes',
        server: 'LiteSpeed',
        connection: 'Keep-Alive' },
     rawHeaders: 
      [ 'Cache-Control',
        'private, no-cache, no-store, must-revalidate, max-age=0',
        'Pragma',
        'no-cache',
        'Content-Type',
        'text/html',
        'Content-Length',
        '1139',
        'Date',
        'Mon, 09 Jul 2018 13:22:52 GMT',
        'Accept-Ranges',
        'bytes',
        'Server',
        'LiteSpeed',
        'Connection',
        'Keep-Alive' ],
     trailers: {},
     rawTrailers: [],
     upgrade: false,
     url: '',
     method: null,
     statusCode: 403,
     statusMessage: 'Forbidden',
     client: 
      Socket {
        connecting: false,
        _hadError: false,
        _handle: [Object],
        _parent: null,
        _host: 'raptureorwrath.net',
        _readableState: [Object],
        readable: true,
        domain: null,
        _events: [Object],
        _eventsCount: 8,
        _maxListeners: undefined,
        _writableState: [Object],
        writable: true,
        allowHalfOpen: false,
        destroyed: false,
        _bytesDispatched: 213,
        _sockname: null,
        _pendingData: null,
        _pendingEncoding: '',
        server: null,
        _server: null,
        parser: null,
        _httpMessage: null,
        read: [Function],
        _consuming: true,
        _idleTimeout: -1,
        _idleNext: null,
        _idlePrev: null,
        _idleStart: 1213959 },
     _consuming: true,
     _dumped: false,
     req: 
      ClientRequest {
        domain: null,
        _events: [Object],
        _eventsCount: 6,
        _maxListeners: undefined,
        output: [],
        outputEncodings: [],
        outputCallbacks: [],
        outputSize: 0,
        writable: true,
        _last: true,
        upgrading: false,
        chunkedEncoding: false,
        shouldKeepAlive: true,
        useChunkedEncodingByDefault: false,
        sendDate: false,
        _removedHeader: {},
        _contentLength: 0,
        _hasBody: true,
        _trailer: '',
        finished: true,
        _headerSent: true,
        socket: [Object],
        connection: [Object],
        _header: 'GET /wp-content/uploads/2013/09/stinging-locusts1-300x240.jpg HTTP/1.1\r\nUser-Agent: wikizimmer/0.1.0 (https://github.com/vadp/zimmer email:vadp.devl@gmail.com)\r\nhost: raptureorwrath.net\r\nConnection: keep-alive\r\n\r\n',
        _headers: [Object],
        _headerNames: [Object],
        _onPendingData: null,
        agent: [Object],
        socketPath: undefined,
        timeout: undefined,
        method: 'GET',
        path: '/wp-content/uploads/2013/09/stinging-locusts1-300x240.jpg',
        _ended: true,
        parser: null,
        timeoutCb: null,
        res: [Circular] },
     request: 
      Request {
        domain: null,
        _events: [Object],
        _eventsCount: 5,
        _maxListeners: undefined,
        encoding: null,
        uri: [Object],
        headers: [Object],
        resolveWithFullResponse: true,
        timeout: 300000,
        forever: true,
        resolve: [Function],
        reject: [Function],
        retries: 0,
        readable: true,
        writable: true,
        _qs: [Object],
        _auth: [Object],
        _oauth: [Object],
        _multipart: [Object],
        _redirect: [Object],
        _tunnel: [Object],
        _rp_resolve: [Function],
        _rp_reject: [Function],
        _rp_promise: [Object],
        _rp_callbackOrig: undefined,
        callback: [Function],
        _rp_options: [Object],
        setHeader: [Function],
        hasHeader: [Function],
        getHeader: [Function],
        removeHeader: [Function],
        method: 'GET',
        localAddress: undefined,
        pool: [Object],
        dests: [],
        __isRequestRequest: true,
        _callback: [Function: RP$callback],
        proxy: null,
        tunnel: false,
        setHost: true,
        originalCookieHeader: undefined,
        _disableCookies: true,
        _jar: undefined,
        port: 80,
        host: 'raptureorwrath.net',
        path: '/wp-content/uploads/2013/09/stinging-locusts1-300x240.jpg',
        httpModule: [Object],
        agentClass: [Object],
        agentOptions: [Object],
        agent: [Object],
        _started: true,
        href: 'http://raptureorwrath.net/wp-content/uploads/2013/09/stinging-locusts1-300x240.jpg',
        req: [Object],
        ntick: true,
        timeoutTimer: null,
        response: [Circular],
        originalHost: 'raptureorwrath.net',
        originalHostHeaderName: 'host',
        responseContent: [Circular],
        _destdata: true,
        _ended: true,
        _callbackCalled: true },
     toJSON: [Function: responseToJSON],
     caseless: Caseless { dict: [Object] },
     read: [Function],
     body: <Buffer 3c 21 44 4f 43 54 59 50 45 20 68 74 6d 6c 3e 0a 3c 68 74 6d 6c 20 73 74 79 6c 65 3d 22 68 65 69 67 68 74 3a 31 30 30 25 22 3e 0a 3c 68 65 61 64 3e 3c ... > } } { url: 'http://raptureorwrath.net/wp-content/uploads/2013/09/stinging-locusts1-300x240.jpg',
  encoding: null,
  uri: null,
  headers: { 'User-Agent': 'wikizimmer/0.1.0 (https://github.com/vadp/zimmer email:vadp.devl@gmail.com)' },
  resolveWithFullResponse: true,
  timeout: 300000,
  forever: true,
  resolve: [Function],
  reject: [Function],
  retries: 0 }

```

**Comment by vss-devel** [comment]
It appears http://raptureorwrath.net accepts requests only from browsers. I'll post a fix a bit later. For starters though you'd replace UserAgent's current value with `'Mozilla/5.0 (X11; Linux x86_64; rv:12.0) Gecko/20100101 Firefox/12.0'`. Then it should be OK for your.

**Comment by fattredd** [comment]
That did help with http://raptureorwrath.net, but later down the line I ran into a 503 error. I realize that this wiki is a mess, and is therefor an exception, but perhaps zimmer should handle all 400 and 500 level errors the same way it deals with 404?

**Comment by vss-devel** [comment]
You'd do that, albeit some data wouldn't be downloaded. So, it's better to try to make some analysis first. 

That's why I made a fatal error. But it might be useful to be able to lift this bar.

**Comment by vss-devel** [comment]
It looks like the issue is eventually fixed in the fd189f534fdf6b4011885bf1f8253469eb87c0d8.

The thing is that this wiki is linking to a lot of external images and many of them are not alive or correct anymore. Actually it was an interesting case.

To get the best result in this case you'd use something like:
```
wikizimmer.js https://www.dandwiki.com --user-agent firefox -r -e 0 -p "hydra-media.cursecdn.com/zelda.gamepedia.com|d1u5p3l4wpay3k.cloudfront.net/zelda_gamepedia_en
```

**Comment by vss-devel** [comment]
Closed
